File: naskah.py
#!/usr/bin/env python3
"""Urai naskah berformat Exact Worksheet Maker jadi soal terstruktur.

Berbeda dari gemini_impor (yang menangani keluaran Gemini apa adanya), modul ini
memahami format penuh: bagian (PG)/(B)/(I)/(E), bobot nilai [n], sub-soal
berjenjang, blok Kunci Jawaban, dan blok Pembahasan. Dipakai untuk memasukkan
soal hasil generate ke bank soal supaya bisa dicari dan disusun ulang tanpa AI.
"""
import re, json

BAGIAN = re.compile(r'^\s*Bagian\s+([^:]{2,60}):\s*\((PG|B|I|E|M|IB)\)\s*$', re.I)
BUTIR = re.compile(r'^\s*(PG|B|I|E|M|IB)(\d{1,3})\.\s*(.*)$')
OPSI = re.compile(r'^\s*\(?([A-Ea-e])\s*[.)]\s*(.+)$')
SUB = re.compile(r'^\s*\((?:([a-h])|((?:i{1,3}|iv|v|vi{1,3})))\)\s*(.*)$')
BOBOT = re.compile(r'\s*\[(\d{1,2})\]\s*$')
KEPALA_KUNCI = re.compile(r'^\s*Kunci\s*Jawaban\s*:?\s*$', re.I)
# Teks bacaan (lembar Bahasa Indonesia / Inggris): baris "Bacaan" sendirian
# sesudah judul lembar, lalu judul teks (opsional), lalu paragraf-paragrafnya,
# sampai bagian soal pertama. Ejaan yang diterima sama dengan wsm/app.js.
KEPALA_BACAAN = re.compile(r'^\s*(?:Reading\s*Passage|Reading\s*Text|Passage|Bacaan|'
                           r'Teks\s*Bacaan|Wacana)\s*(?:\d{1,2})?\s*:?\s*$', re.I)
# Batas akhir bacaan: kepala bagian apa pun (termasuk "Bagian A: (PG)" yang
# namanya satu huruf — BAGIAN di atas minta 2 huruf) atau butir soal pertama.
BATAS_BACAAN = re.compile(r'^\s*(?:Bagian|Section)\s+[^\n]{1,80}?:\s*(?:\([A-Za-z]{1,4}\))?\s*$', re.I)
KEPALA_BAHAS = re.compile(r'^\s*Pembahasan\s*:?\s*$', re.I)
SET = re.compile(r'^\s*SET\s+(\d+)\s*$', re.I)
# "PG1-B, I1-18, E1a-5x+2" — kode, tanda hubung, lalu isinya sampai PENANDA
# KODE BERIKUTNYA (atau habis). Dulu isinya dibatasi koma: cocok untuk kunci
# yang pendek, tapi blok Pembahasan dari Gemini datang tanpa pemisah sama
# sekali ("E1a-$...$E1b-$...$") dan kalimatnya penuh koma — hasilnya
# pembahasan terpotong di koma pertama, sisanya hilang, dan Exact Practice
# menerima "0 berpembahasan". Pemecah ini disamakan dengan pbRe/akRe milik
# Worksheet Maker (wsm/app.js), jadi PDF dan Practice membaca blok yang sama
# dengan cara yang sama. Koma pemisah di akhir isi dibuang di _pasang().
_KODE = r'(?:PG|IB|B|I|E|M)\d{1,3}[a-h]?(?:\.(?:i{1,3}|iv|v|vi{1,3}))?'
PASANG = re.compile(r'(?<![A-Za-z])(' + _KODE + r')\s*-\s*([\s\S]*?)(?=\s*,?\s*(?<![A-Za-z])'
                    + _KODE + r'\s*-|\s*$)')

def _pasang(teks):
    """{KODE: isi} dari blok Kunci Jawaban / Pembahasan."""
    return {k.upper(): _bersih(v).rstrip(',').strip() for k, v in PASANG.findall(teks)}

JENIS = {'PG': 'Pilihan Ganda', 'B': 'Benar/Salah', 'I': 'Isian Singkat',
         'E': 'Uraian', 'M': 'Menjodohkan', 'IB': 'Isian Berpilihan'}

def _bersih(t):
    return re.sub(r'\s{2,}', ' ', (t or '').strip())

def urai(teks):
    """Kembalikan (daftar_soal, meta). Tiap soal punya kode, jenis, batang,
    opsi, bobot, sub-soal, dan kunci bila ada.

    Naskah bisa memuat beberapa set ("SET 1", "SET 2", …), masing-masing dengan
    blok Kunci Jawaban dan Pembahasan sendiri. Dulu seluruh naskah dibaca
    sebagai satu aliran: begitu "Kunci Jawaban" set 1 lewat, semua baris set 2
    dianggap pembahasan — soal set 2 hilang, dan kuncinya (kode yang sama,
    PG1…) menimpa kunci set 1. Sekarang tiap set diurai sendiri-sendiri.
    """
    baris = teks.replace('\r', '').split('\n')
    potongan, kini_set, kini_baris = [], 1, []
    for b in baris:
        m = SET.match(b)
        if m:
            potongan.append((kini_set, kini_baris)); kini_set, kini_baris = int(m.group(1)), []
        else:
            kini_baris.append(b)
    potongan.append((kini_set, kini_baris))
    potongan = [(n, br) for n, br in potongan if any(x.strip() for x in br)] or [(1, baris)]
    soal, judul, n_kunci = [], '', 0
    for n, br in potongan:
        s, j, k = _urai_set(br, n)
        soal += s; n_kunci += k; judul = judul or j
    bacaan = next((s.get('bacaan') for s in soal if s.get('bacaan')), None)
    return soal, {'judul': judul, 'n_kunci': n_kunci, 'n_set': max([s['set'] for s in soal] or [1]),
                  'bacaan': bacaan}

def _ambil_bacaan(isi):
    """Pisahkan blok teks bacaan dari baris-baris soal.

    Kembalikan (isi_tanpa_bacaan, bacaan|None); bacaan = {'judul', 'isi'}.
    Judul teks = paragraf pertama kalau pendek, satu baris, dan masih ada
    paragraf lain sesudahnya — aturan yang sama dengan wsm/app.js supaya PDF
    dan Practice membaca judul yang sama.
    """
    awal = next((i for i, b in enumerate(isi) if KEPALA_BACAAN.match(b)), None)
    if awal is None: return isi, None
    batas = next((i for i, b in enumerate(isi) if BATAS_BACAAN.match(b) or BUTIR.match(b)), len(isi))
    if awal >= batas: return isi, None
    blok = '\n'.join(isi[awal + 1:batas]).strip()
    para = [p.strip() for p in re.split(r'\n[ \t]*\n', blok) if p.strip()]
    judul = ''
    if len(para) > 1 and len(para[0]) <= 120 and '\n' not in para[0]:
        judul = para.pop(0)
    if not para: return isi[:awal] + isi[batas:], None
    return isi[:awal] + isi[batas:], {'judul': judul, 'isi': '\n\n'.join(para)}


def _urai_set(baris, set_ini=1):
    """Urai SATU set: soal, lalu blok Kunci Jawaban dan Pembahasan miliknya."""
    bagian_kunci, bagian_bahas = [], []
    isi, mode = [], 'soal'
    for b in baris:
        if KEPALA_KUNCI.match(b): mode = 'kunci'; continue
        if KEPALA_BAHAS.match(b): mode = 'bahas'; continue
        (isi if mode == 'soal' else bagian_kunci if mode == 'kunci' else bagian_bahas).append(b)
    isi, bacaan = _ambil_bacaan(isi)
    # Judul lembar = baris pertama yang berisi, dibaca SESUDAH bacaan dipisah
    # — kalau tidak, lembar tanpa judul yang langsung "Bacaan" berjudul "Bacaan".
    judul = next((b.strip() for b in isi[:4] if b.strip()), '')
    if BATAS_BACAAN.match(judul) or BUTIR.match(judul): judul = ''   # naskah tanpa judul

    kunci = _pasang('\n'.join(bagian_kunci))
    bahas = _pasang('\n'.join(bagian_bahas))

    soal, kini, huruf = [], None, None
    def tutup():
        nonlocal kini
        if kini:
            kini['batang'] = _bersih(kini['batang'])
            if len(kini['batang']) >= 5: soal.append(kini)
        kini = None

    for b in isi:
        if BAGIAN.match(b): tutup(); continue
        m = BUTIR.match(b)
        if m:
            tutup()
            sisa, bobot = m.group(3), None
            mb = BOBOT.search(sisa)
            if mb: bobot = int(mb.group(1)); sisa = sisa[:mb.start()]
            kini = {'kode': f'{m.group(1).upper()}{m.group(2)}', 'jenis': m.group(1).upper(),
                    'no': int(m.group(2)), 'batang': sisa, 'opsi': {}, 'bobot': bobot,
                    'sub': [], 'set': set_ini}
            huruf = None
            continue
        if kini is None: continue
        m = SUB.match(b)
        if m:
            sisa, bobot = m.group(3), None
            mb = BOBOT.search(sisa)
            if mb: bobot = int(mb.group(1)); sisa = sisa[:mb.start()]
            if m.group(1):
                huruf = m.group(1)
                kini['sub'].append({'label': huruf, 'teks': _bersih(sisa), 'bobot': bobot})
            else:
                kini['sub'].append({'label': f'{huruf or "a"}.{m.group(2)}',
                                    'teks': _bersih(sisa), 'bobot': bobot})
            continue
        m = OPSI.match(b)
        if m and kini['jenis'] in ('PG', 'IB'):
            kini['opsi'][m.group(1).upper()] = _bersih(m.group(2)); continue
        if kini['sub']: kini['sub'][-1]['teks'] += ' ' + b.strip()
        elif kini['opsi']:
            akhir = sorted(kini['opsi'])[-1]; kini['opsi'][akhir] += ' ' + b.strip()
        else: kini['batang'] += ' ' + b.strip()
    tutup()

    for s in soal:
        s['kunci'] = kunci.get(s['kode'], '')
        s['pembahasan'] = bahas.get(s['kode'], '')
        # Bacaan menempel di tiap soal (bukan hanya di meta): Exact Practice
        # menyimpan soal satu per satu sebagai stimulus, dan bank soal juga.
        s['bacaan'] = bacaan
        for sub in s['sub']:
            k = (s['kode'] + sub['label']).upper()
            if k in kunci: sub['kunci'] = kunci[k]
            if k in bahas: sub['pembahasan'] = bahas[k]
        # Soal uraian berbagian: kunci dan pembahasannya per bagian (E1a-, E1b-).
        # Induknya ikut diisi gabungannya supaya tetap terbaca "berkunci" /
        # "berpembahasan" dan ikut ke Practice/bank soal sebagai satu teks.
        if not s['kunci'] and any(sub.get('kunci') for sub in s['sub']):
            s['kunci'] = '; '.join(f"({sub['label']}) {sub['kunci']}" for sub in s['sub'] if sub.get('kunci'))
        if not s['pembahasan'] and any(sub.get('pembahasan') for sub in s['sub']):
            s['pembahasan'] = '\n'.join(f"({sub['label']}) {sub['pembahasan']}" for sub in s['sub'] if sub.get('pembahasan'))
    return soal, judul, len(kunci)

def ke_naskah(daftar, judul='LEMBAR KERJA'):
    """Kebalikannya: susun soal terpilih jadi naskah siap ditempel ke
    Exact Worksheet Maker — inilah yang membuat penyusunan ulang tanpa AI."""
    per_jenis = {}
    for s in daftar: per_jenis.setdefault(s.get('jenis') or 'PG', []).append(s)
    baris = [judul, '']
    # Teks bacaan ikut ditulis ulang supaya paket dari Practice / bank soal
    # yang dicetak ulang tetap membawa bacaannya. Satu bacaan per naskah:
    # kalau soal-soalnya membawa bacaan berbeda, yang pertama yang dipakai.
    bacaan = next((s.get('bacaan') for s in daftar
                   if isinstance(s.get('bacaan'), dict) and (s['bacaan'].get('isi') or '').strip()), None)
    if bacaan:
        baris.append('Bacaan')
        # Judul teks dan paragraf pertama dipisah baris kosong — begitulah
        # pengurai (dan wsm/app.js) membedakan judul dari isi.
        if (bacaan.get('judul') or '').strip(): baris += [bacaan['judul'].strip(), '']
        baris += [bacaan['isi'].strip(), '']
    kunci = []
    bahas = []
    for kode in ('PG', 'B', 'I', 'E', 'M', 'IB'):
        if kode not in per_jenis: continue
        baris.append(f'Bagian {JENIS[kode]}: ({kode})')
        for i, s in enumerate(per_jenis[kode], 1):
            bobot = f" [{s['bobot']}]" if s.get('bobot') else ''
            baris.append(f"{kode}{i}. {s['batang']}{bobot}")
            for h, v in sorted((s.get('opsi') or {}).items()):
                baris.append(f'{h}. {v}')
            for sub in (s.get('sub') or []):
                b2 = f" [{sub['bobot']}]" if sub.get('bobot') else ''
                baris.append(f"({sub['label']}) {sub['teks']}{b2}")
            if s.get('kunci'): kunci.append(f"{kode}{i}-{s['kunci']}")
            if s.get('pembahasan'): bahas.append(f"{kode}{i}-{s['pembahasan']}")
        baris.append('')
    if kunci:
        baris += ['Kunci Jawaban', ', '.join(kunci)]
    # Pembahasan dulu hilang di sini: soal membawa pembahasannya, tapi naskah
    # yang disusun tidak pernah menuliskannya, jadi PDF "kunci & pembahasan"
    # yang dicetak lewat /api/render keluar tanpa satu pun pembahasan. Satu
    # entri per baris -- pemecah di Worksheet Maker mencari penanda kode
    # berikutnya, bukan koma, jadi kalimat pembahasan boleh memuat koma.
    if bahas:
        baris += ['', 'Pembahasan'] + bahas
    return '\n'.join(baris)

File: src/lib/practice/naskah.ts
/* Pengurai naskah berformat Exact Worksheet Maker — port TypeScript dari
 * naskah.py (urai()) milik Exact Worksheet.
 *
 * Kenapa diport, bukan memanggil Worksheet lewat HTTP seperti jembatan lain di
 * worksheet.ts: justru inti alur "tempel dari AI" adalah tidak bergantung pada
 * mesin Gemini/Chrome di Worksheet — guru menempel naskah dari ChatGPT/Gemini/
 * Claude mana pun, dan itu harus tetap jalan saat Worksheet mati. Selain itu
 * berkas ini murni teks (tanpa modul node), jadi pengurai yang sama dipakai di
 * peramban untuk pratinjau langsung DAN di server saat menyimpan — pratinjau
 * tidak mungkin beda dengan hasil akhirnya.
 *
 * Format yang dipahami sama persis dengan yang dikeluarkan Worksheet Maker:
 *
 *   Bagian Pilihan Ganda: (PG)
 *   PG1. Teks soal [2]
 *   A. opsi pertama
 *   B. opsi kedua
 *
 *   Kunci Jawaban
 *   PG1-B, I1-18
 *
 *   Pembahasan
 *   PG1-langkahnya, I1-langkahnya
 *
 * Beberapa set dipisah baris "SET 2" dan diurai sendiri-sendiri, supaya kunci
 * set 2 (kodenya berulang dari PG1 lagi) tidak menimpa kunci set 1.
 */
import type { Bacaan, Butir } from "./worksheet";

const BAGIAN = /^\s*Bagian\s+([^:]{2,60}):\s*\((PG|B|I|E|M|IB)\)\s*$/i;
const BUTIR = /^\s*(PG|B|I|E|M|IB)(\d{1,3})\.\s*(.*)$/;
const OPSI = /^\s*\(?([A-Ea-e])\s*[.)]\s*(.+)$/;
const SUB = /^\s*\((?:([a-h])|(i{1,3}|iv|v|vi{1,3}))\)\s*(.*)$/;
const BOBOT = /\s*\[(\d{1,2})\]\s*$/;
const KEPALA_KUNCI = /^\s*Kunci\s*Jawaban\s*:?\s*$/i;
/* Teks bacaan (lembar Bahasa Indonesia / Inggris): baris "Bacaan" sendirian
 * di bawah judul naskah, lalu judul teks (opsional), lalu paragraf-paragrafnya
 * sampai bagian soal pertama. Ejaan yang diterima sama dengan naskah.py dan
 * wsm/app.js milik Exact Worksheet. */
const KEPALA_BACAAN = /^\s*(?:Reading\s*Passage|Reading\s*Text|Passage|Bacaan|Teks\s*Bacaan|Wacana)\s*(?:\d{1,2})?\s*:?\s*$/i;
/* Batas akhir bacaan: kepala bagian apa pun (termasuk "Bagian A: (PG)" yang
 * namanya satu huruf — BAGIAN di atas minta 2 huruf) atau butir soal pertama. */
const BATAS_BACAAN = /^\s*(?:Bagian|Section)\s+[^\n]{1,80}?:\s*(?:\([A-Za-z]{1,4}\))?\s*$/i;
const KEPALA_BAHAS = /^\s*Pembahasan\s*:?\s*$/i;
const TANDA_SET = /^\s*SET\s+(\d+)\s*$/i;
/* "PG1-B, I1-18, E1a-5x+2" — kode, tanda hubung, lalu jawaban sampai koma yang
 * diikuti kode berikutnya. Jawaban sendiri boleh memuat koma (mis. "1, 2, 3"),
 * jadi batasnya dicari lewat lookahead, bukan dengan memecah di tiap koma. */
const PASANG = /\b((?:PG|B|I|E|M|IB)\d{1,3}[a-h]?(?:\.(?:i{1,3}|iv|v|vi{1,3}))?)\s*-\s*([^,]+?)(?=\s*,\s*(?:PG|B|I|E|M|IB)\d|\s*$)/gs;

export interface MetaNaskah {
  judul: string;
  /** teks bacaan set pertama, kalau naskahnya lembar pemahaman bacaan */
  bacaan?: Bacaan | null;
  /** banyaknya entri di blok Kunci Jawaban (semua set digabung) */
  nKunci: number;
  nSet: number;
}

/* Penanda kode di awal baris, untuk blok yang ditulis satu entri per baris. */
const AWAL_BARIS = /^\s*((?:PG|B|I|E|M|IB)\d{1,3}[a-h]?(?:\.(?:i{1,3}|iv|v|vi{1,3}))?)\s*-\s*(.*)$/;
const HITUNG_KODE = /\b(?:PG|B|I|E|M|IB)\d{1,3}[a-h]?(?:\.(?:i{1,3}|iv|v|vi{1,3}))?\s*-/g;

const bersih = (t: string) => (t || "").trim().replace(/\s{2,}/g, " ");

function pasanganKoma(teks: string): Record<string, string> {
  const out: Record<string, string> = {};
  for (const m of teks.matchAll(PASANG)) out[m[1].toUpperCase()] = bersih(m[2]);
  return out;
}

function pasanganBaris(baris: string[]): Record<string, string> {
  const out: Record<string, string> = {};
  let kode: string | null = null;
  for (const b of baris) {
    const m = AWAL_BARIS.exec(b);
    if (m) { kode = m[1].toUpperCase(); out[kode] = bersih(m[2]); }
    else if (kode && b.trim()) out[kode] += " " + b.trim();
  }
  return out;
}

/* Blok kunci/pembahasan datang dalam dua bentuk, dan keduanya harus terbaca:
 *
 *   a) satu baris berisi daftar dipisah koma — "PG1-B, PG2-C" (bentuk yang
 *      dikeluarkan Exact Worksheet Maker sendiri lewat ke_naskah());
 *   b) satu entri per baris — bentuk yang diminta prompt kita untuk blok
 *      Pembahasan, karena kalimat penjelasan hampir selalu memuat koma, dan
 *      di bentuk (a) koma itu memotong penjelasannya di tengah jalan.
 *
 * Pembedanya: di bentuk (b) tiap baris hanya memuat SATU kode. Begitu ada satu
 * baris dengan dua kode atau lebih, baris itu pasti daftar berkoma, jadi
 * seluruh blok dibaca sebagai bentuk (a) — kalau tidak, "PG1-B, PG2-C" akan
 * terbaca sebagai satu entri PG1 bernilai "B, PG2-C". */
function pasanganLuwes(baris: string[]): Record<string, string> {
  const satuKodePerBaris = baris
    .filter((b) => b.trim())
    .every((b) => (b.match(HITUNG_KODE) ?? []).length <= 1);
  return satuKodePerBaris ? pasanganBaris(baris) : pasanganKoma(baris.join("\n"));
}

/* Pisahkan blok teks bacaan dari baris-baris soal. Judul teks = paragraf
 * pertama kalau pendek, satu baris, dan masih ada paragraf lain sesudahnya —
 * aturan yang sama dengan naskah.py / wsm/app.js supaya PDF dan Practice
 * membaca judul yang sama. */
function ambilBacaan(isi: string[]): { sisa: string[]; bacaan: Bacaan | null } {
  const awal = isi.findIndex((b) => KEPALA_BACAAN.test(b));
  if (awal < 0) return { sisa: isi, bacaan: null };
  let batas = isi.findIndex((b) => BATAS_BACAAN.test(b) || BUTIR.test(b));
  if (batas < 0) batas = isi.length;
  if (awal >= batas) return { sisa: isi, bacaan: null };
  const blok = isi.slice(awal + 1, batas).join("\n").trim();
  const para = blok.split(/\n[ \t]*\n/).map((p) => p.trim()).filter(Boolean);
  let judul = "";
  if (para.length > 1 && para[0].length <= 120 && !/\n/.test(para[0])) judul = para.shift()!;
  const sisa = [...isi.slice(0, awal), ...isi.slice(batas)];
  return { sisa, bacaan: para.length ? { judul, isi: para.join("\n\n") } : null };
}

/** Urai SATU set: soalnya, lalu blok Kunci Jawaban dan Pembahasan miliknya. */
function uraiSet(baris: string[], setIni: number): { soal: Butir[]; judul: string; nKunci: number; bacaan: Bacaan | null } {
  const isiMentah: string[] = [], barisKunci: string[] = [], barisBahas: string[] = [];
  let mode: "soal" | "kunci" | "bahas" = "soal";
  for (const b of baris) {
    if (KEPALA_KUNCI.test(b)) { mode = "kunci"; continue; }
    if (KEPALA_BAHAS.test(b)) { mode = "bahas"; continue; }
    (mode === "soal" ? isiMentah : mode === "kunci" ? barisKunci : barisBahas).push(b);
  }
  const { sisa: isi, bacaan } = ambilBacaan(isiMentah);
  /* Judul dibaca SESUDAH bacaan dipisah — kalau tidak, naskah tanpa judul
   * yang langsung "Bacaan" berjudul "Bacaan". */
  const judulMentah = isi.slice(0, 4).map((b) => b.trim()).find(Boolean) ?? "";
  const judul = BATAS_BACAAN.test(judulMentah) || BUTIR.test(judulMentah) ? "" : judulMentah;

  const kunci = pasanganLuwes(barisKunci);
  const bahas = pasanganLuwes(barisBahas);

  const soal: Butir[] = [];
  let kini: Butir | null = null;
  let huruf: string | null = null;

  /* Batang di bawah 5 karakter hampir pasti sisa salah-urai (mis. baris "PG3."
   * yang teksnya kebawa ke baris berikutnya), bukan soal sungguhan. */
  const tutup = () => {
    if (kini) {
      kini.batang = bersih(kini.batang);
      if (kini.batang.length >= 5) soal.push(kini);
    }
    kini = null;
  };

  for (const b of isi) {
    if (BAGIAN.test(b)) { tutup(); continue; }

    const mButir = BUTIR.exec(b);
    if (mButir) {
      tutup();
      let sisa = mButir[3];
      let bobot: number | null = null;
      const mb = BOBOT.exec(sisa);
      if (mb) { bobot = Number(mb[1]); sisa = sisa.slice(0, mb.index); }
      kini = {
        kode: `${mButir[1].toUpperCase()}${mButir[2]}`, jenis: mButir[1].toUpperCase(),
        no: Number(mButir[2]), batang: sisa, opsi: {}, bobot, sub: [], set: setIni,
      };
      huruf = null;
      continue;
    }
    if (!kini) continue;

    const mSub = SUB.exec(b);
    if (mSub) {
      let sisa = mSub[3];
      let bobot: number | null = null;
      const mb = BOBOT.exec(sisa);
      if (mb) { bobot = Number(mb[1]); sisa = sisa.slice(0, mb.index); }
      if (mSub[1]) {
        huruf = mSub[1];
        kini.sub!.push({ label: huruf, teks: bersih(sisa), bobot });
      } else {
        kini.sub!.push({ label: `${huruf ?? "a"}.${mSub[2]}`, teks: bersih(sisa), bobot });
      }
      continue;
    }

    const mOpsi = OPSI.exec(b);
    if (mOpsi && (kini.jenis === "PG" || kini.jenis === "IB")) {
      kini.opsi[mOpsi[1].toUpperCase()] = bersih(mOpsi[2]);
      continue;
    }

    /* Baris lanjutan: disambung ke bagian terakhir yang sedang dibaca, supaya
     * soal/opsi yang terbungkus beberapa baris tidak terpotong. */
    if (kini.sub!.length) kini.sub![kini.sub!.length - 1].teks += " " + b.trim();
    else if (Object.keys(kini.opsi).length) {
      const akhir = Object.keys(kini.opsi).sort().pop()!;
      kini.opsi[akhir] += " " + b.trim();
    } else kini.batang += " " + b.trim();
  }
  tutup();

  for (const s of soal) {
    s.kunci = kunci[s.kode] ?? "";
    s.pembahasan = bahas[s.kode] ?? "";
    /* Bacaan menempel di tiap soal: soal disimpan satu per satu sebagai
     * Question, dan tiap Question membawa stimulusnya sendiri. */
    s.bacaan = bacaan;
  }
  return { soal, judul, nKunci: Object.keys(kunci).length, bacaan };
}

/** Urai naskah lengkap (boleh berisi beberapa SET) jadi butir soal + meta. */
export function uraiNaskah(teks: string): { butir: Butir[]; meta: MetaNaskah } {
  const baris = (teks || "").replace(/\r/g, "").split("\n");

  const potongan: { set: number; baris: string[] }[] = [];
  let kiniSet = 1, kiniBaris: string[] = [];
  for (const b of baris) {
    const m = TANDA_SET.exec(b);
    if (m) { potongan.push({ set: kiniSet, baris: kiniBaris }); kiniSet = Number(m[1]); kiniBaris = []; }
    else kiniBaris.push(b);
  }
  potongan.push({ set: kiniSet, baris: kiniBaris });

  const terpakai = potongan.filter((p) => p.baris.some((x) => x.trim()));
  const dipakai = terpakai.length ? terpakai : [{ set: 1, baris }];

  const butir: Butir[] = [];
  let judul = "", nKunci = 0;
  let bacaan: Bacaan | null = null;
  for (const p of dipakai) {
    const h = uraiSet(p.baris, p.set);
    butir.push(...h.soal);
    nKunci += h.nKunci;
    judul = judul || h.judul;
    bacaan = bacaan ?? h.bacaan;
  }
  return { butir, meta: { judul, bacaan, nKunci, nSet: Math.max(1, ...butir.map((b) => b.set ?? 1)) } };
}
